File: MATLAB/triangle.m
function [waveforms, noteCycle, phaseIncrement] = triangle(noteFreq, fs, velocity)
% ---------------------------------------------------------------
% Generate a single-cycle triangle wave using precalculated wavetable banks
% ---------------------------------------------------------------
%
% This function creates a single-cycle triangle waveform for a given note frequency
% using precomputed wavetable banks. Only odd harmonics are included with amplitudes
% scaled according to the triangle wave formula. The waveform is phase-aware and
% ready for continuous playback.
%
% Inputs:
%   noteFreq        - Frequency of the note to play (Hz)
%   fs              - Sampling rate (Hz)
%   velocity        - Amplitude scaling (0 to 1)
%
% Outputs:
%   waveforms       - Cell array of all precalculated wavetable banks
%   noteCycle       - Single-cycle waveform for the given note frequency
%   phaseIncrement  - Amount to increment phase per sample (for continuous playback)
%
% Notes:
%   - The function limits harmonics per bank to avoid aliasing.
%   - Only odd harmonics are used to synthesize the triangle wave, with alternating signs.
%   - The note frequency is constrained to the available bank range.
% ---------------------------------------------------------------

% Central frequencies for the banks
banks = [20, 40, 80, 160, 320, 640, 1280, 5120];

% Nyquist frequency
fNyquist = fs / 2;

% Maximum number of desired harmonics per bank
desired_number_of_harmonics = 64;

% Initialize cell array to store normalized waveforms
waveforms = cell(1, length(banks));

% Generate wavetable banks
for b = 1:length(banks)
    bankFreq = banks(b);

    % Number of samples per period
    Ns = round(fs / bankFreq);
    n = 0:Ns-1;

    % Maximum harmonics allowed by Nyquist
    maxHarmonics = floor(fNyquist / bankFreq);
    number_of_harmonics = min(desired_number_of_harmonics, maxHarmonics);

    % Harmonic vector and amplitudes (triangle wave formula, odd harmonics)
    k = 1:number_of_harmonics;
    a = mod(k,2) ./ (k.^2);       % odd harmonics only
    a = a .* ((-1).^( (k-1)/2 )); % alternate sign for triangle wave

    % Initialize waveform
    waveform = zeros(1, Ns);

    % Sum harmonics
    for harmonic_index = 1:number_of_harmonics
        waveform = waveform + a(harmonic_index) * sin(2*pi*harmonic_index*n/Ns);
    end

    % Normalize waveform
    waveform_normalized = waveform / max(abs(waveform));

    % Store in cell array
    waveforms{b} = waveform_normalized;
end

% -------------------
% Select closest bank
% -------------------
% Limit noteFreq to bank range to avoid out-of-range errors
noteFreq = max(min(noteFreq, banks(end)), banks(1));
[~, bankIndex] = min(abs(banks - noteFreq));
bankWave = waveforms{bankIndex};

% Interpolation to match exact note frequency (single cycle)
Ns_new = round(fs / noteFreq);
t_old = linspace(0, 1, length(bankWave));
t_new = linspace(0, 1, Ns_new);
noteCycle = interp1(t_old, bankWave, t_new, 'linear');

% Normalize and apply velocity
noteCycle = velocity * (noteCycle / max(abs(noteCycle)));

% Compute phase increment for continuous playback
phaseIncrement = noteFreq / fs;

end

File: Plugin/src/Oscillator/Waves/Triangle.cpp
#include "Triangle.h"
#include <cmath>
#include <algorithm>

/**
 * @brief Constructs a Triangle oscillator and initializes its wavetable banks.
 * @param sampleRate Sampling rate in Hz.
 * @param velocity Amplitude scaling factor.
 *
 * Populates the internal list of frequency banks and generates a series of
 * band-limited triangle waveforms for efficient playback.
 */
Triangle::Triangle(double sampleRate, float velocity)
    : WavetableOscillator({}, {}), banks({20, 25, 30, 35, 40, 45, 50, 56, 63, 70, 79, 88, 99, 111, 125,
                                          140, 157, 176, 198, 222, 249, 279, 313, 352, 395, 443, 497, 558,
                                          627, 704, 790, 886, 992, 1111, 1244, 1393, 1560, 1747, 1955, 2187,
                                          2450, 2740, 3070, 3440, 3860, 4330, 4860, 5450, 6120, 6870, 7720,
                                          8680, 9760, 10960, 12320, 13840, 15530, 17440, 19550, 20000})
{
    name = "Triangle";
    generateWavetableBanks(sampleRate, velocity);
}

/**
 * @brief Builds a series of band-limited triangle waveforms.
 * @param fs Sampling rate in Hz.
 * @param velocity Amplitude scaling factor.
 *
 * Each waveform is constructed using a truncated Fourier series containing
 * only odd harmonics, following:
 *
 * \f[
 * x(t) = \sum_{n=0}^{N} (-1)^n \frac{1}{(2n+1)^2} \sin(2 \pi (2n+1) t)
 * \f]
 *
 * The number of harmonics is limited by both a fixed maximum and the
 * Nyquist constraint. After generation, each waveform is normalized to
 * unity and scaled by the velocity factor.
 */
void Triangle::generateWavetableBanks(double fs, float velocity)
{
    waveTables.clear();
    waveTables.reserve(banks.size());

    const int desiredNumberOfHarmonics = 128;
    double fNyquist = fs / 2.0;

    for (auto bankFreq : banks)
    {
        int Ns = static_cast<int>(std::round(fs / bankFreq));
        std::vector<float> waveform(Ns, 0.0f);

        int numHarmonics = (bankFreq > 10000.0) ? 32 : desiredNumberOfHarmonics;
        int maxHarmonics = static_cast<int>(0.9 * fNyquist / bankFreq);
        numHarmonics = std::min(numHarmonics, maxHarmonics);

        // odd harmonics with alternating sign and 1/n^2 amplitude
        for (int n = 0; n < numHarmonics; ++n)
        {
            int k = 2 * n + 1;
            float amplitude = std::pow(-1.0f, n) / (k * k);
            for (int i = 0; i < Ns; ++i)
                waveform[i] += amplitude * std::sin(2.0 * M_PI * k * i / Ns);
        }

        // normalize
        float maxVal = *std::max_element(waveform.begin(), waveform.end(),
                                         [](float a, float b)
                                         { return std::abs(a) < std::abs(b); });
        for (auto &v : waveform)
            v = velocity * v / maxVal;

        waveTables.push_back(waveform);
    }

    bankFreqs = banks;
}
